File: examples/Clojure/mtrelay.clj
(ns mtrelay
  (:refer-clojure :exclude [send])
  (:require [zhelpers :as mq])
  (:import [org.zeromq ZMQ$Context]))

;;
;; Multithreaded relay in Clojure
;; Isaiah Peng <issaria@gmail.com>
;;

(defrecord Step1 [^ZMQ$Context ctx]
  Runnable
  (run [this]
    (let [xmitter (mq/socket ctx mq/pair)]
      (mq/connect xmitter "inproc://step2")
      (println "Step 1 is ready, signaling step 2")
      (mq/send xmitter "READY\u0000")
      (.close xmitter))))

(defrecord Step2 [^ZMQ$Context ctx]
  Runnable
  (run [this]
    (let [receiver (mq/socket ctx mq/pair)
          xmitter (mq/socket ctx mq/pair)]
      (mq/bind receiver "inproc://step2")
      (-> ctx Step1. Thread. .start)
      ;; Wait for signal and pass it on
      (mq/recv-str receiver)
      (mq/connect xmitter "inproc://step3")
      (println "Step 2 is ready, signaling step 3")
      (mq/send xmitter "READY\u0000")
      (.close xmitter))))

(defn -main []
  (let [ctx (mq/context 1)
        receiver (mq/socket ctx mq/pair)]
    (mq/bind receiver "inproc://step3")
    (-> ctx Step2. Thread. .start)
    (mq/recv-str receiver)
    (.close receiver)
    (println "Test successful!")
    (.term ctx)))

File: examples/Rust/mtrelay.rs
#![crate_name = "mtrelay"]

extern crate zmq;

use std::thread;

fn step1(context: &zmq::Context) {
    let xmitter = context.socket(zmq::PAIR).unwrap();
    assert!(xmitter.connect("inproc://step2").is_ok());
    println!("Step 1 ready, signaling step 2");
    xmitter.send_str("READY", 0).unwrap();
}

fn step2(context: &zmq::Context) {
    let receiver = context.socket(zmq::PAIR).unwrap();
    assert!(receiver.bind("inproc://step2").is_ok());

    let ncontext = context.clone();
    thread::spawn(move || step1(&ncontext));

    let _ = receiver.recv_string(0).unwrap().unwrap();

    let xmitter = context.socket(zmq::PAIR).unwrap();
    assert!(xmitter.connect("inproc://step3").is_ok());
    println!("Step 2 ready, signaling step 3");
    xmitter.send_str("READY", 0).unwrap();
}

fn main() {
    let context = zmq::Context::new();

    let receiver = context.socket(zmq::PAIR).unwrap();
    assert!(receiver.bind("inproc://step3").is_ok());

    let ncontext = context.clone();
    thread::spawn(move || step2(&ncontext));

    let _ = receiver.recv_string(0).unwrap().unwrap();

    println!("Test successful!");
}
